include/rdma/fabric.h

```diff
@@ -465,6 +481,14 @@ struct fi_fabric_attr {
 	uint32_t		api_version;
 };
 
+struct fi_hmem_attr {
+	enum fi_hmem_iface		iface;
+	enum fi_hmem_attr_opt		api_permitted;
```

Comment:
This is acting as a bool.

Reply:
These fields will be both input and output. Since a boolean can represent either false or an unspecified value, FI_HMEM_ATTR_UNSPEC is added to differentiate them.

Reply:
Either the provider can call the GPU API or not.  There's not a third state here.

These settings are dictating to the provider how it must *implement* data transfers to/from GPU buffers.  For some providers, it means that the provider cannot support GPU buffers at all.  (There is no PCI peer to peer support for TCP or shmem.)  There' a significant difference between these variables and other attribute values.

Reply:
>it means that the provider cannot support GPU buffers at all. (There is no PCI peer to peer support for TCP or shmem.) 

I think that is part of our goal for making it in fi_info because it can help the provider filtering. For provider that doesn't support PCIe peer-to-peer, like shmem, we can use such configuration to filter it early in fi_info. You remember we have had challenges to toggle shmem on/off inside EFA provider without using environment variables. You believed shm usage is a data transfer level decision so making it as a ep level setopt() makes more sense. The option can be either a general `FI_OPT_SHARED_MEMORY_PERMITTED` or `FI_OPT_HMEM_P2P`

We currently use `FI_OPT_SHARED_MEMORY_PERMITTED`, but such toggle in ep level is still too late for us because we have created shm info/domain/av/cq/mr earlier. Cleaning all of them at an ep call is troublesome and also error-prone.

Making such toggle as early as in fi_info level can resolve such challenge.

>There' a significant difference between these variables and other attribute values.

I agree on this point. If possible, we can consider alternatives to move them to appropriate attribute groups (like domain/tx/rx/ep_attr) separately.

Reply:
I agree having this information up front is useful.  I disagree that these are extended attributes.  They're something else.  Environment variables are the closest thing to what this intends to capture.

Imagine 2 upper libraries (e.g. MPI and NCCL) calling libfabric.  These libraries could drive a provider in different directions.  NCCL says "no, don't use CUDA", but MPI says "go ahead and use CUDA".  That doesn't work.  There are *global* settings at play here, not per domain or endpoint settings.  These settings may not even be per provider. NCCL might be using ucx, but MPI verbs, yet the restrictions from NCCL need to carry over both providers.

Reply:
> These libraries could drive a provider in different directions. NCCL says "no, don't use CUDA", but MPI says "go ahead and use CUDA". That doesn't work. 

This is exactly the problem we are solving. In the consideration of resource management and other factors, making Libfabric use cuda calls in both control and data transfer interfaces for NCCL application may cause unexpected risks and overhead. But we don't have such concern for MPI application.

>NCCL might be using ucx, but MPI verbs, yet the restrictions from NCCL need to carry over both providers.

Can you explain to me the software stack here? I know NCCL can use UCX/OFI for network offload via the plugins. It can also use MPI (via a NCCL-MPI plugin ?) for the same purpose?

I learned NCCL is already warning users that using NCCL and CUDA aware MPI together is not safe: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/mpi.html#inter-gpu-communication-with-cuda-aware-mpi

Reply:
I'm referring to a single process using BOTH NCCL and MPI.  Or a single process accessing libfabric through more than 1 middleware (NCCL and DAOS, MPI and DAOS, etc.).  The point is that these settings aren't per domain or per provider but are global to the process.  That is, they are settings which apply to the _environment_ as a whole.

Reply:
NCCL doesn't like users to use CUDA aware MPI calls in the same progress. So I don't see a reason we cannot make MPI allow libfabric to use CUDA while NCCL doesn't allow.